### ISSUE TITLE
Can't update tags of mp3 file with chapters markers

### ISSUE THREAD

**Issue by LucaTNT** [JamesHeinrich/getID3#186]
Hello,
I've been using getID3 with great success for editing tags and adding album art pictures, however I had never tried to work on mp3 files that already had chapter markers in them.

If I try to edit any tag in a file that has chapter markers in it, then I loose all chapter markers.

Any hints would be appreciated. If necessary, I can supply code snippets and example mp3s.

**Comment by JamesHeinrich** [comment]
Writing functions are still considered beta and not necessarily feature-complete.

If you provide a code snippet and sample file I can take a quick look to see where the problem might be.

**Comment by LucaTNT** [comment]
Thank you, @JamesHeinrich!

Here is a code snippet, sample file and sample cover art: [getid3_chapters_sample.zip](https://github.com/JamesHeinrich/getID3/files/3204502/getid3_chapters_sample.zip)

**Comment by JamesHeinrich** [comment]
I took a quick look, a few comments:

1) You're trying to write ID3v2.2 which is not supported. Please use ID3v2.3 instead.

2) Your chapters are disappearing because you're using `$tagwriter->overwrite_tags = true;` which is defined as `If true will erase existing tag data and write only passed data`. Of course, you're also using that because if you set it to true you'll get the fatal warning that it's currently not supported (per my above comments about the writing code being incomplete).

In short, what you need to do as a workaround is to read in the current tag data, pick out the bits you want to keep and add it to the data you're writing to rewrite it all back in the new tag. This is not easy or convenient, I know.

What really needs to be done is the whole tag-writing module(s) need to be fundamentally rewritten, which is a major undertaking that I don't currently have the resources for.

**Comment by LucaTNT** [comment]
I uploaded the wrong php file, this is the correct one, and I also added code to try and copy chapter data, to no avail.

```php
<?php
require 'getid3/getid3.php';
copy('edit_me.mp3', 'edited.mp3');


$getID3 = new getID3;
$getID3->setOption(array('encoding' => 'UTF-8'));
$old_tags = $getID3->analyze('edited.mp3');
getid3_lib::IncludeDependency(GETID3_INCLUDEPATH.'write.php', __FILE__, true);
$tagwriter = new getid3_writetags;
$tagwriter->filename       = 'edited.mp3';
$tagwriter->tagformats     = array( 'id3v2.3' );
$tagwriter->overwrite_tags    = true;
$tagwriter->remove_other_tags = false;

$TagData = array(
	'title'    => array('This is my title'),
	'artist'   => array("This is my artist"),
	'album'    => array("This is my album"),
	'year'     => array(2019),
	'genre'    => array("This is my genre"),
	'track'    => array(12),
	'composer' => array("This is my composer"),
	'CTOC'     => $old_tags['id3v2']['CTOC'],
	'CHAP'     => $old_tags['id3v2']['CHAP'],
	'chapters' => $old_tags['id3v2']['chapters']
);

$fd = fopen('cover.jpg', 'rb');
$APICdata = fread($fd, filesize('cover.jpg'));
fclose($fd);

$TagData['attached_picture'][0]['data']          = $APICdata;
$TagData['attached_picture'][0]['picturetypeid'] = 3;
$TagData['attached_picture'][0]['description']   = 'Cover';
$TagData['attached_picture'][0]['mime']          = 'image/jpeg';


$tagwriter->tag_data = $TagData;

if(!$tagwriter->WriteTags())
{
	die("Error: ".implode("\n", $tagwriter->errors));
}

if (!empty($tagwriter->warnings)) {
	echo "There were some warnings:\n".implode("\n", $tagwriter->warnings);
}

print_r($getID3->analyze('edited.mp3'));
```

Do you have any further hint as how copy chapters over?
Thank you very much for all the work you've done, I really appreciate it.

**Comment by yknivag** [comment]
Would also like to be able to write CTOC and CHAP tags. Is this at all possible at the current time?

**Comment by BJTheCod3r** [comment]
> I took a quick look, a few comments:
> 
>     1. You're trying to write ID3v2.2 which is not supported. Please use ID3v2.3 instead.
> 
>     2. Your chapters are disappearing because you're using `$tagwriter->overwrite_tags = true;` which is defined as `If true will erase existing tag data and write only passed data`. Of course, you're also using that because if you set it to true you'll get the fatal warning that it's currently not supported (per my above comments about the writing code being incomplete).
> 
> 
> In short, what you need to do as a workaround is to read in the current tag data, pick out the bits you want to keep and add it to the data you're writing to rewrite it all back in the new tag. This is not easy or convenient, I know.
> 
> What really needs to be done is the whole tag-writing module(s) need to be fundamentally rewritten, which is a major undertaking that I don't currently have the resources for.

I'm actually having related issues as @LucaTNT. I'd seen this as the only way but was just hoping for a miracle. Thanks for the clarification, at least I would stop feeling like I'm doing it wrong and make the best of the situation.

**Comment by shealan** [comment]
Hey all, did you manage to come up with a way of creating new chapters with titles, urls and images? I'd love to see an example if you did :)

**Comment by machadoug** [comment]
@JamesHeinrich and @StudioMaX , can you please provide an example on how to write chapter tags? I tried to find examples here and check the documentation on id3.org, but id3.org is offline (500 error).

I'm trying to add the following array to the chapters key, but it's not working. All other tags are working properly. Also, where to add Chapter images? 

```php
$getID3 = new getID3;
$getID3->setOption(['encoding' => 'UTF-8']);

$writer                 = new getid3_writetags;
$writer->tagformats     = ['id3v2.3'];
$writer->overwrite_tags = true;
$writer->tag_encoding   = 'UTF-8';


$tagData			= [
	'title'  => [$item->title],
	'artist' => ['My Artist'],
	'album'  => [$item->category_title],
	'track'  => [$item->episode],
	'year'   => [$date->format('Y')],
	'date'   => [$date->format('Y-m-d')],
	'time'   => [$date->format('H:i')],
	'composer'   => [$app->get('sitename')],
	'chapters' => []
];

$tagData['comment'] = array(strip_tags(HTMLHelper::_('content.prepare', $item->description)));

$item->outline = array_values($item->outline);
$nChapters = count($item->outline);

for ($i = 0; $i < $nChapters; $i++)
{
	$chapter = $item->outline[$i];

	if (($i + 1) < $nChapters)
	{
		$endTime = $item->outline[$i + 1]['time'];
	}
	else
	{
		$endTime = $item->duration;
	}

	$chapterTag = [
		'time_begin'	=> EpisodeHelper::timeToSeconds($chapter['time']),
		'time_end'		=> EpisodeHelper::timeToSeconds($endTime),
		'chapter_name'	=> $chapter['heading'] ?? ($chapter['description'] ?? '')
	];

	if (!empty($chapter['url']))
	{
		$chapterTag['chapter_url'] = [$chapter['url']];
	}

	$tagData['chapters'][] = $chapterTag;
}

$writer->tag_data = $tagData;

$writer->filename = 'my/path';
$writer->WriteTags();
```

Thank you in advance
